src/Processors/Formats/IOutputFormat.h

```diff
@@ -85,6 +95,55 @@ class IOutputFormat : public IProcessor
         }
     }
 
+    void writeSuffixIfNot()
+    {
+        if (need_write_suffix)
+        {
+            writeSuffix();
+            need_write_suffix = false;
+        }
+    }
+
+    /// Methods-helpers for parallel formatting.
+
+    /// Set the real number of first row in the first chunk of data.
+    void setFirstRowNumber(size_t first_row_number_)
+    {
+        first_row_number = first_row_number_;
+        onFirstRowNumberUpdate();
+    }
+
+    size_t getFirstRowNumber() const { return first_row_number; }
+
+    /// Update state according to new first_row_number.
+    virtual void onFirstRowNumberUpdate() {}
+
+    /// Some formats outputs some statistics after the data,
+    /// in parallel formatting we collect these statistics outside the
+    /// underling format and then set it to format before finalizing.
+    struct Statistics
+    {
+        Stopwatch watch;
+        Progress progress;
+        bool applied_limit = false;
+        size_t rows_before_limit = 0;
+        Chunk totals;
+        Chunk extremes;
+    };
+
+    void setOutsideStatistics(Statistics statistics_) { statistics = std::make_shared<Statistics>(std::move(statistics_)); }
+    std::shared_ptr<Statistics> getOutsideStatistics() const { return statistics; }
+
+    /// In some formats the way we print extremes depends on
+    /// were totals printed or not. In this case in parallel formatting
+    /// we should notify underling format if totals were printed.
+    void setTotalsAreWritten() { are_totals_written = true; }
+    bool areTotalsWritten() const { return are_totals_written; }
+
+    /// Return true if format saves totals and extremes in consumeTotals/consumeExtremes and
+    /// outputs them in finalize() method.
+    virtual bool areTotalsAndExtremesUsedInFinalize() const { return false; }
```

Comment:
Why do we need this? Aren't your code generic enough? Underlying format may decide for itself what to do with the statistics?

Reply:
UPD: I understood

Reply:
The main problem is that in parallel formatting we use separate threads for consumeTotals/consumeExtremes/finalize (and it's not bad), but some formats can save some state in these methods and use it in the other method. I haven't come up with anything better than just save such states in ParallelFormattingOutputFormat and set it to formatters in threads. It made code a bit tricky and complicated, but I thought it worth it (and maybe we will try to refactor it in future).